File: frontend/src/types/event.ts
/**
 * Event Types
 *
 * This file contains all type definitions related to events.
 * Events represent incidents and maintenance periods in the system.
 *
 * All types follow JavaScript naming conventions (camelCase).
 * API response conversion happens in the API service layer.
 */

/**
 * Event kinds (matches backend Kind enum)
 */
export type EventKind = "incident" | "maintenance";

/**
 * Event data structure (JavaScript conventions)
 *
 * This is the canonical event type used throughout the application.
 * API services convert backend responses to this format.
 */
export interface Event {
	/** Unique identifier for the event */
	id: number;
	/** Event title/name */
	title: string;
	/** Optional description providing details about the event */
	description: string | null;
	/** Type of event (incident or maintenance) */
	kind: EventKind;
	/** Key-value labels associated with this event */
	labels: Record<string, string>;
	/** When the event started (ISO datetime string) */
	startedAt: string;
	/** When the event ended (ISO datetime string) */
	endedAt: string;
}

/**
 * Input type for creating a new event
 * Used when posting to the API
 */
export interface EventCreateInput {
	/** Event title/name */
	title: string;
	/** Optional description providing details about the event */
	description?: string | null;
	/** Type of event (incident or maintenance) */
	kind: EventKind;
	/** Key-value labels associated with this event */
	labels?: Record<string, string>;
	/** When the event started (ISO datetime string) */
	startedAt: string;
	/** When the event ended (ISO datetime string) */
	endedAt: string;
}

/**
 * Store structure for efficient event lookups
 * Maps event IDs to their data
 */
export interface EventsStore {
	[eventId: number]: Event;
}

File: backend/src/models/event.py
"""Event models and entities."""

from datetime import datetime
from enum import Enum

from pydantic import BaseModel
from sqlalchemy import JSON
from sqlalchemy.orm import Mapped, mapped_column

from .base import Base


class EventKind(str, Enum):
    """Event type enumeration."""

    incident = "incident"
    maintenance = "maintenance"


class EventEntity(BaseModel):
    """Pydantic entity for Event."""

    id: int
    title: str
    description: str | None = None
    kind: EventKind
    labels: dict[str, str] = {}
    started_at: datetime
    ended_at: datetime


class Event(Base):
    """SQLAlchemy model for events."""

    __tablename__ = "event"

    uuid: Mapped[str] = mapped_column(primary_key=True)
    title: Mapped[str]
    description: Mapped[str | None]
    kind: Mapped[str]
    labels: Mapped[dict[str, str]] = mapped_column(JSON)
    created_at: Mapped[datetime] = mapped_column(default=datetime.now)
    started_at: Mapped[datetime]
    ended_at: Mapped[datetime]

    def to_entity(self) -> EventEntity:
        """Convert SQLAlchemy model to Pydantic entity."""
        return EventEntity(
            id=self.id,
            title=self.title,
            description=self.description,
            kind=EventKind(self.kind),
            labels=self.labels,
            started_at=self.started_at,
            ended_at=self.ended_at,
        )
